components/tabs/src/storage.rs

```diff
@@ -362,6 +372,34 @@ impl TabsStorage {
     }
 }
 
+// Trim the amount of tabs in a list to fit the specified memory size
+fn trim_tabs_length(
+    tabs: &mut Vec<RemoteTab>,
+    payload_size_max_bytes: usize,
+) -> Result<Vec<RemoteTab>> {
+    // Same as what desktop does in util.js
+    // See bug 535326 comment 8 for an explanation of the estimation
+    let max_serialized_size = (payload_size_max_bytes / 4) * 3 - 1500;
+    let size = compute_serialized_size(&tabs);
+    if size > max_serialized_size {
+        // Estimate a little more than the direct fraction to maximize packing
+        let cutoff = (tabs.len() * max_serialized_size) / size;
+        log::debug!("Truncating the tabs by: {}", cutoff);
+        tabs.truncate(cutoff);
+
+        // Keep dropping off the last entry until the data fits.
+        while compute_serialized_size(&tabs) > max_serialized_size {
+            tabs.pop();
+        }
+    }
+    Ok(tabs.to_vec())
+}
+
+fn compute_serialized_size(v: &Vec<RemoteTab>) -> usize {
+    let json = serde_json::to_string(v).unwrap();
```

Comment:
This unwrap is wrong (even though I don't think it will ever fail tbh, but everywhere else handles serde failures so we should too). It looks like this is the only reason the calling functions need a `Result<>`